pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConsumerHashAssignmentsSnapshot.java

```diff
@@ -0,0 +1,206 @@
+            mergedRangesByConsumer.put(consumer, RemovedHashRanges.of(mergeOverlappingRanges(ranges)));
+        });
+        return ImpactedConsumersResult.of(mergedRangesByConsumer);
+    }
+
+    static SortedSet<Range> mergeOverlappingRanges(SortedSet<Range> ranges) {
+        TreeSet<Range> mergedRanges = new TreeSet<>();
+        Iterator<Range> rangeIterator = ranges.iterator();
+        Range currentRange = rangeIterator.hasNext() ? rangeIterator.next() : null;
+        while (rangeIterator.hasNext()) {
+            Range nextRange = rangeIterator.next();
+            if (currentRange.getEnd() >= nextRange.getStart() - 1) {
+                currentRange = Range.of(currentRange.getStart(), Math.max(currentRange.getEnd(), nextRange.getEnd()));
+            } else {
+                mergedRanges.add(currentRange);
+                currentRange = nextRange;
+            }
+        }
+        if (currentRange != null) {
+            mergedRanges.add(currentRange);
+        }
+        return mergedRanges;
+    }
+
+    /**
+     * Calculate the diff of two range mappings.
+     * @param mappingBefore the range mapping before
+     * @param mappingAfter the range mapping after
+     * @return the impacted ranges where the consumer is changed from the before to the after
+     */
+    static Map<Range, Pair<Consumer, Consumer>> diffRanges(SortedMap<Range, Consumer> mappingBefore,
+                                                           SortedMap<Range, Consumer> mappingAfter) {
+        Map<Range, Pair<Consumer, Consumer>> removedRanges = new LinkedHashMap<>();
+        Iterator<Map.Entry<Range, Consumer>> beforeIterator = mappingBefore.entrySet().iterator();
+        Iterator<Map.Entry<Range, Consumer>> afterIterator = mappingAfter.entrySet().iterator();
+
+        Map.Entry<Range, Consumer> beforeEntry = beforeIterator.hasNext() ? beforeIterator.next() : null;
+        Map.Entry<Range, Consumer> afterEntry = afterIterator.hasNext() ? afterIterator.next() : null;
+
+        while (beforeEntry != null && afterEntry != null) {
+            Range beforeRange = beforeEntry.getKey();
+            Range afterRange = afterEntry.getKey();
+            Consumer beforeConsumer = beforeEntry.getValue();
+            Consumer afterConsumer = afterEntry.getValue();
+
+            if (beforeRange.equals(afterRange)) {
+                if (!beforeConsumer.equals(afterConsumer)) {
+                    removedRanges.put(afterRange, Pair.of(beforeConsumer, afterConsumer));
+                }
+                beforeEntry = beforeIterator.hasNext() ? beforeIterator.next() : null;
+                afterEntry = afterIterator.hasNext() ? afterIterator.next() : null;
+            } else if (beforeRange.getEnd() < afterRange.getStart()) {
+                removedRanges.put(beforeRange, Pair.of(beforeConsumer, afterConsumer));
+                beforeEntry = beforeIterator.hasNext() ? beforeIterator.next() : null;
```

Comment:
> it's implemented in this method so that there's a generic diff method if we later on need to also find out which ranges were added. Since this won't be a performance bottleneck, I don't see a need to optimize only for the removed ranges.

I agree, but I asked for just confirmation.
Thanks.